src/libsyntax/tokenstream.rs

```diff
@@ -115,14 +115,15 @@ impl TokenTree {
         let open_span = if span.is_dummy() {
             span
         } else {
+            assert!(span.lo() != span.hi());
             span.with_hi(span.lo() + BytePos(delim.len() as u32))
         };
         TokenTree::token(token::OpenDelim(delim), open_span)
     }
 
     /// Returns the closing delimiter as a token tree.
     pub fn close_tt(span: Span, delim: DelimToken) -> TokenTree {
-        let close_span = if span.is_dummy() {
+        let close_span = if span.is_dummy() || span.lo() == span.hi() {
             span
         } else {
             span.with_lo(span.hi() - BytePos(delim.len() as u32))
```

Comment:
This may be trickier than it may look on the first glance.
If `span` is arbitrary (e.g. created by a proc macro), then `span + 1` can always step into something bad in the source map.
The only way to know for sure is to check whether `span_to_source` returns an error or not, but that's expensive.

If we want to check this without accessing the source map, then we need to understand when exactly this span is used for indexing into it (only when reporting "un-closed delimiter" or somewhere else?) and whether it can contain spans produced by proc macros at that point, or only lexer produced spans.

Reply:
the tokenstream would insert an un-closed delimiter with a span whose `lo` equals `hi`. I'm not sure about proc macros, haven't got time into that part yet.

Reply:
@petrochenkov I think it would be reasonable to commit this PR with a FIXME comment pretty much containing the same text in your comment, as well as a ticket for that work. Although this won't catch every single ICE that could be caused by proc_macro span shenaningans, it will catch one that is relatively easy to hit. We could also add a check for whether the span has a macro backtrace and also use `span` directly in those cases.

Reply:
I'd rather add some asserts instead of the FIXME, it's quite possible that this is only called with spans from lexer, will try to look into this tomorrow.

I'd also inline `open_tt` and `close_tt` into the caller since they are only called once and cannot work with arbitrary arguments.